File: src/DataScience/MATLAB/composite.m
function composite
    readme = fileNode(2);
    docs = folderNode({fileNode(3), fileNode(5)});
    root = folderNode({readme, docs});

    fprintf('leaf=%d\n', nodeSize(readme));
    fprintf('docs=%d\n', nodeSize(docs));
    fprintf('root=%d\n', nodeSize(root));
end

function node = fileNode(bytes)
    node = struct('kind', "file", 'bytes', bytes, 'children', {{}});
end

function node = folderNode(children)
    node = struct('kind', "folder", 'bytes', 0, 'children', {children});
end

function total = nodeSize(node)
    if node.kind == "file"
        total = node.bytes;
        return;
    end

    total = 0;
    for index = 1:numel(node.children)
        total = total + nodeSize(node.children{index});
    end
end

File: src/Niche/Nim/composite.nim
type
  NodeKind = enum
    nkFile, nkFolder

  Node = ref object
    case kind: NodeKind
    of nkFile:
      bytes: int
    of nkFolder:
      children: seq[Node]

proc fileNode(bytes: int): Node =
  Node(kind: nkFile, bytes: bytes)

proc folderNode(children: varargs[Node]): Node =
  Node(kind: nkFolder, children: @children)

proc size(node: Node): int =
  case node.kind
  of nkFile:
    result = node.bytes
  of nkFolder:
    for child in node.children:
      result += child.size()

let readme = fileNode(2)
let docs = folderNode(fileNode(3), fileNode(5))
let root = folderNode(readme, docs)

echo "leaf=", readme.size()
echo "docs=", docs.size()
echo "root=", root.size()
